File: plugin/src/knowledge/providers/dbpedia.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const DBPEDIA_SPARQL = "https://dbpedia.org/sparql";
const DBPEDIA_LOOKUP = "https://lookup.dbpedia.org/api/search";

interface DBpediaLookupResult {
  docs?: Array<{
    resource?: string[];
    label?: string[];
    comment?: string[];
    category?: string[];
    type?: string[];
  }>;
}

interface SparqlResult {
  results?: {
    bindings?: Array<{
      [key: string]: { value: string; type?: string };
    }>;
  };
}

export class DBpediaProvider implements KnowledgeProvider {
  readonly name = "dbpedia";

  async isAvailable(): Promise<boolean> {
    try {
      const response = await requestUrl({
        url: `${DBPEDIA_LOOKUP}?query=test&maxResults=1&format=json`,
        method: "GET",
        throw: false,
      });
      return response.status === 200;
    } catch {
      return false;
    }
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    try {
      // Use lookup API for search
      const entries = await this.searchLookup(query, maxResults);

      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `DBpedia lookup failed: ${message}`,
      };
    }
  }

  private async searchLookup(query: string, maxResults: number): Promise<KnowledgeEntry[]> {
    const params = new URLSearchParams({
      query,
      maxResults: String(maxResults),
      format: "json",
    });

    const response = await requestUrl({
      url: `${DBPEDIA_LOOKUP}?${params}`,
      method: "GET",
      headers: {
        Accept: "application/json",
      },
      throw: false,
    });

    if (response.status !== 200) {
      throw new Error(`Lookup request failed: ${response.status}`);
    }

    const data = response.json as DBpediaLookupResult;

    if (!data.docs || !Array.isArray(data.docs)) {
      return [];
    }

    return data.docs
      .filter((doc) => doc.label?.[0])
      .map((doc) => {
        const resource = doc.resource?.[0] ?? "";
        const types = doc.type?.slice(0, 3).map((t) => t.split("/").pop()) ?? [];

        return {
          title: doc.label?.[0] ?? "Unknown",
          summary: doc.comment?.[0] ?? "",
          url: resource,
          source: "dbpedia",
          metadata: {
            types,
            categories: doc.category?.slice(0, 5),
            resourceUri: resource,
          },
        };
      });
  }

  async getResourceDetails(resourceUri: string): Promise<KnowledgeEntry | null> {
    const sparqlQuery = `
      SELECT ?label ?abstract ?thumbnail ?type WHERE {
        <${resourceUri}> rdfs:label ?label .
        OPTIONAL { <${resourceUri}> dbo:abstract ?abstract . FILTER(lang(?abstract) = "en") }
        OPTIONAL { <${resourceUri}> dbo:thumbnail ?thumbnail }
        OPTIONAL { <${resourceUri}> rdf:type ?type . FILTER(STRSTARTS(STR(?type), "http://dbpedia.org/ontology/")) }
        FILTER(lang(?label) = "en")
      }
      LIMIT 1
    `;

    const params = new URLSearchParams({
      query: sparqlQuery,
      format: "json",
    });

    const response = await requestUrl({
      url: `${DBPEDIA_SPARQL}?${params}`,
      method: "GET",
      headers: {
        Accept: "application/sparql-results+json",
      },
      throw: false,
    });

    if (response.status !== 200) {
      return null;
    }

    const data = response.json as SparqlResult;
    const binding = data.results?.bindings?.[0];

    if (!binding?.label?.value) {
      return null;
    }

    return {
      title: binding.label.value,
      summary: binding.abstract?.value?.slice(0, 500) ?? "",
      url: resourceUri,
      source: "dbpedia",
      metadata: {
        type: binding.type?.value?.split("/").pop(),
        thumbnail: binding.thumbnail?.value,
      },
    };
  }
}

File: crates/mcp/src/knowledge/dbpedia.rs
use reqwest::blocking::Client;
use serde::Deserialize;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const DBPEDIA_LOOKUP: &str = "https://lookup.dbpedia.org/api/search";

pub struct DBpediaProvider {
    client: Client,
}

#[derive(Debug, Deserialize)]
struct LookupResponse {
    docs: Option<Vec<Doc>>,
}

#[derive(Debug, Deserialize)]
struct Doc {
    resource: Option<Vec<String>>,
    label: Option<Vec<String>>,
    comment: Option<Vec<String>>,
    category: Option<Vec<String>>,
    #[serde(rename = "type")]
    types: Option<Vec<String>>,
}

impl DBpediaProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
        }
    }
}

impl Default for DBpediaProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for DBpediaProvider {
    fn name(&self) -> &'static str {
        "dbpedia"
    }

    fn is_available(&self) -> bool {
        self.client
            .get(&format!("{}?query=test&maxResults=1&format=json", DBPEDIA_LOOKUP))
            .send()
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }

    fn lookup(&self, query: &str, options: &LookupOptions) -> LookupResult {
        let limit = options.max_results.unwrap_or(5);

        let url = format!(
            "{}?query={}&maxResults={}&format=json",
            DBPEDIA_LOOKUP,
            urlencoding::encode(query),
            limit
        );

        let response = match self.client.get(&url).header("Accept", "application/json").send() {
            Ok(r) => r,
            Err(e) => return LookupResult::error(self.name(), e.to_string()),
        };

        if !response.status().is_success() {
            return LookupResult::error(
                self.name(),
                format!("lookup request failed: {}", response.status()),
            );
        }

        let data: LookupResponse = match response.json() {
            Ok(d) => d,
            Err(e) => return LookupResult::error(self.name(), e.to_string()),
        };

        let entries: Vec<KnowledgeEntry> = data
            .docs
            .unwrap_or_default()
            .into_iter()
            .filter_map(|doc| {
                let label = doc.label.as_ref()?.first()?.clone();
                let resource = doc.resource.as_ref().and_then(|r| r.first().cloned());
                let comment = doc.comment.as_ref().and_then(|c| c.first().cloned());

                let types: Vec<String> = doc
                    .types
                    .unwrap_or_default()
                    .into_iter()
                    .take(3)
                    .filter_map(|t| t.rsplit('/').next().map(String::from))
                    .collect();

                let categories: Vec<String> = doc
                    .category
                    .unwrap_or_default()
                    .into_iter()
                    .take(5)
                    .collect();

                let mut metadata = std::collections::HashMap::new();
                if !types.is_empty() {
                    metadata.insert("types".to_string(), serde_json::json!(types));
                }
                if !categories.is_empty() {
                    metadata.insert("categories".to_string(), serde_json::json!(categories));
                }
                if let Some(ref uri) = resource {
                    metadata.insert("resourceUri".to_string(), serde_json::json!(uri));
                }

                Some(KnowledgeEntry {
                    title: label,
                    summary: comment.unwrap_or_default(),
                    url: resource,
                    source: "dbpedia".to_string(),
                    metadata: if metadata.is_empty() {
                        None
                    } else {
                        Some(metadata)
                    },
                })
            })
            .collect();

        LookupResult::success(self.name(), entries)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore] // Requires network
    fn dbpedia_lookup() {
        let provider = DBpediaProvider::new();
        let result = provider.lookup("Rust programming language", &LookupOptions::default());
        assert!(result.success);
        assert!(!result.entries.is_empty());
    }
}
